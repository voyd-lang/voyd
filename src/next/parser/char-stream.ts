import { SourceLocation } from "./ast/syntax.js";

export class CharStream {
  readonly filePath: string;
  readonly value: string;
  readonly length: number;
  readonly location = {
    index: 0,
    line: 1,
    column: 0,
  };

  constructor(contents: string, filePath: string) {
    this.value = contents;
    this.length = contents.length;
    this.filePath = filePath;
  }

  /** Current index the file is on */
  get position() {
    return this.location.index;
  }

  get line() {
    return this.location.line;
  }

  get column() {
    return this.location.column;
  }

  get hasCharacters() {
    return this.position < this.length;
  }

  get next() {
    return this.value[this.position];
  }

  currentSourceLocation() {
    const startLine = this.line;
    const startColumn = this.column;
    return new SourceLocation({
      startIndex: this.position,
      endIndex: this.position,
      startLine,
      endLine: startLine,
      startColumn,
      endColumn: startColumn,
      filePath: this.filePath,
    });
  }

  at(index: number): string | undefined {
    return this.value[this.position + index];
  }

  /** Returns the next character and advances the stream */
  consumeChar(): string {
    const char = this.next;
    if (char === undefined) {
      throw new Error("Out of characters");
    }

    this.location.index += 1;
    this.location.column += 1;
    if (char === "\n") {
      this.location.line += 1;
      this.location.column = 0;
    }

    return char;
  }
}
