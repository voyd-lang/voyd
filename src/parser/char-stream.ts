import { SourceLocation } from "../syntax-objects/syntax.js";

export class CharStream {
  readonly filePath: string;
  readonly originalSize: number;
  readonly contents: string[];
  readonly location = {
    index: 0,
    line: 1,
    column: 0,
  };

  constructor(contents: string, filePath: string) {
    this.contents = contents.split("");
    this.originalSize = this.contents.length;
    this.filePath = filePath;
  }

  /** Current index the file is on */
  get position() {
    return this.location.index;
  }

  get line() {
    return this.location.line;
  }

  get column() {
    return this.location.column;
  }

  get hasCharacters() {
    return this.position < this.originalSize;
  }

  get next() {
    return this.contents[this.position];
  }

  currentSourceLocation() {
    return new SourceLocation({
      startIndex: this.position,
      endIndex: this.position,
      line: this.line,
      column: this.column,
      filePath: this.filePath,
    });
  }

  at(index: number): string | undefined {
    return this.contents[this.position + index];
  }

  /** Returns the next character and removes it from the queue */
  consumeChar(): string {
    const char = this.contents[this.position];
    if (char === undefined) {
      throw new Error("Out of characters");
    }

    this.location.index += 1;
    this.location.column += 1;
    if (char === "\n") {
      this.location.line += 1;
      this.location.column = 0;
    }

    return char;
  }
}
