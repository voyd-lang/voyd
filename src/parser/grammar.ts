import { Expr } from "../syntax-objects/expr.js";
import { Identifier } from "../syntax-objects/identifier.js";

export const idIs = (id: Expr | undefined, value: string) =>
  id?.isIdentifier() && id.is(value);

export const isTerminator = (char: string) =>
  isWhitespace(char) ||
  isBracket(char) ||
  isQuote(char) ||
  isOpChar(char) ||
  char === ",";

export const isQuote = newTest(["'", '"', "`"]);

export const isWhitespace = (char: string) =>
  char === " " || char === "\n" || char === "\r";

export const isBracket = newTest(["{", "[", "(", ")", "]", "}"]);

export const isOpChar = newTest([
  "+",
  "-",
  "*",
  "/",
  "=",
  ":",
  "?",
  ".",
  ";",
  "<",
  ">",
  "$",
  "!",
  "@",
  "%",
  "^",
  "&",
  "~",
  "\\",
  "#",
]);

export const isDigit = (char: string) => char >= "0" && char <= "9";
export const isDigitSign = (char: string) => char === "+" || char === "-";

/** Key is the operator, value is its precedence */
export type OpMap = Map<string, number>;

export const infixOps: OpMap = new Map([
  ["+", 1],
  ["-", 1],
  ["*", 2],
  ["/", 2],
  ["^", 3],
  ["%", 2],
  ["==", 0],
  ["!=", 0],
  ["<", 1],
  [">", 1],
  ["<=", 1],
  [">=", 1],
  [".", 6],
  ["|>", 4],
  ["<|", 4],
  ["|", 4],
  ["=", 0],
  ["+=", 4],
  ["-=", 4],
  ["*=", 4],
  ["/=", 4],
  ["->", 5],
  ["=>", 5],
  [":", 0],
  ["?:", 0],
  ["?=", 0],
  ["::", 6],
  [";", 4],
  ["??", 3],
  ["?.", 6],
  ["and", 0],
  ["or", 0],
  ["xor", 0],
  ["as", 0],
  ["is", 0],
  ["is_subtype_of", 0],
  ["in", 0],
  ["has_trait", 0],
]);

export const isInfixOp = (op?: Expr): op is Identifier =>
  !!op?.isIdentifier() && isInfixOpIdentifier(op);

export const isInfixOpIdentifier = (op?: Identifier) =>
  !!op && !op.isQuoted && infixOps.has(op.value);

export const isOp = (op?: Expr): boolean => isInfixOp(op) || isPrefixOp(op);

export const prefixOps: OpMap = new Map([
  ["#", 0],
  ["&", 7],
  ["!", 7],
  ["~", 7],
  ["%", 7],
  ["$", 7],
  ["@", 7],
  ["$@", 7],
  ["...", 5],
]);

export const isPrefixOp = (op?: Expr): op is Identifier =>
  !!op?.isIdentifier() && isPrefixOpIdentifier(op);

export const isPrefixOpIdentifier = (op?: Identifier) =>
  !!op && !op.isQuoted && prefixOps.has(op.value);

export const greedyOps = new Set(["=>", "=", "<|", ";", "|"]);

export const isGreedyOp = (expr?: Expr): expr is Identifier => {
  if (!expr?.isIdentifier()) return false;
  return isGreedyOpIdentifier(expr);
};

export const isGreedyOpIdentifier = (op?: Identifier) =>
  !!op && !op.isQuoted && greedyOps.has(op.value);

export const isContinuationOp = (op?: Expr) => isInfixOp(op);

function newTest<T>(list: Set<T> | Array<T>) {
  const set = new Set(list);
  return (val: T) => set.has(val);
}
