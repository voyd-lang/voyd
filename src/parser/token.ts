import { SourceLocation } from "../syntax-objects/syntax.js";

const WHITESPACE = /^\s+$/;

export class Token {
  readonly location: SourceLocation;
  value = "";

  constructor(opts: { location: SourceLocation; value?: string }) {
    const { value, location } = opts;
    this.value = value ?? "";
    this.location = location;
  }

  get length() {
    return this.value.length;
  }

  get hasChars() {
    return !!this.value.length;
  }

  get isNumber() {
    return /^[0-9]+$/.test(this.value);
  }

  get first(): string | undefined {
    return this.value[0];
  }

  get isWhitespace() {
    return WHITESPACE.test(this.value);
  }

  addChar(string: string) {
    this.value += string;
  }

  is(string?: string) {
    return this.value === string;
  }

  setEndLocationToStartOf(location?: SourceLocation) {
    this.location.setEndToStartOf(location);
  }
}
