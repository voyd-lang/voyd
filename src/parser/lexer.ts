import { Token } from "./token.js";
import { CharStream } from "./char-stream.js";
import {
  isOpChar,
  isTerminator,
  isWhitespace,
  isDigit,
  isDigitSign,
} from "./grammar.js";

const NUMBER_REGEX = /^[+-]?\d+(?:\.\d+)?([Ee]?[+-]?\d+|(?:i|f)(?:|3|6|32|64))?$/;

/**
 * Lexer that tracks angle bracket nesting depth so that `>>` can be
 * tokenized either as two closing brackets (inside generics) or as a shift
 * operator (outside generics).
 */
export class Lexer {
  private angleBracketDepth = 0;

  tokenize(chars: CharStream): Token {
    const token = new Token({
      location: chars.currentSourceLocation(),
    });

    while (chars.hasCharacters) {
      const char = chars.next;

      if (!token.hasChars && char === " ") {
        this.consumeSpaces(chars, token);
        break;
      }

      if (!token.hasChars && this.nextIsNumber(chars)) {
        this.consumeNumber(chars, token);
        break;
      }

      if (!token.hasChars && char === ",") {
        token.addChar(chars.consumeChar());
        break;
      }

      if (!token.hasChars && isOpChar(char)) {
        this.consumeOperator(chars, token);
        break;
      }

      if (!token.hasChars && isTerminator(char)) {
        token.addChar(chars.consumeChar());
        break;
      }

      // Support sharp identifiers (Used by reader macros ignores non-whitespace terminators)
      if (token.first === "#" && !isWhitespace(char)) {
        token.addChar(chars.consumeChar());
        continue;
      }

      if (char === "\t") {
        throw new Error(
          "Tabs are not supported, use four spaces for indentation",
        );
      }

      if (isTerminator(char)) {
        break;
      }

      token.addChar(chars.consumeChar());
    }

    token.setEndLocationToStartOf(chars.currentSourceLocation());

    this.updateAngleBracketDepth(token, chars);

    return token;
  }

  private consumeOperator(chars: CharStream, token: Token) {
    while (isOpChar(chars.next)) {
      if (token.value === ">" && (this.angleBracketDepth > 0 || chars.next === ":")) {
        break;
      }

      token.addChar(chars.consumeChar());
    }
  }

  private updateAngleBracketDepth(token: Token, chars: CharStream) {
    if (token.value === "<" && !isWhitespace(chars.next)) {
      this.angleBracketDepth += 1;
    } else if (token.value === ">" && this.angleBracketDepth > 0) {
      this.angleBracketDepth -= 1;
    }
  }

  private consumeNumber(chars: CharStream, token: Token) {
    while (true) {
      const next = chars.next;
      if (
        !next ||
        (!NUMBER_REGEX.test(token.value + next) &&
          !NUMBER_REGEX.test(token.value + next + (chars.at(1) ?? "")))
      ) {
        break;
      }
      token.addChar(chars.consumeChar());
    }
  }

  private nextIsNumber(chars: CharStream) {
    return (
      isDigit(chars.next) ||
      (isDigitSign(chars.next) && isDigit(chars.at(1) ?? ""))
    );
  }

  private consumeSpaces(chars: CharStream, token: Token) {
    while (chars.next === " " && token.length < 2) {
      token.addChar(chars.consumeChar());
    }
  }
}

