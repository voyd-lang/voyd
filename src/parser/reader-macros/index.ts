import { Token } from "../token.js";
import { arrayLiteralMacro } from "./array-literal.js";
import { booleanMacro } from "./boolean.js";
import { comment } from "./comment.js";
import { mapLiteralMacro } from "./map-literal.js";
import { floatMacro } from "./float.js";
import { intMacro } from "./int.js";
import { scientificENotationMacro } from "./scientific-e-notation.js";
import { stringMacro } from "./string.js";
import { objectLiteralMacro } from "./object-literal.js";
import { ReaderMacro } from "./types.js";
import { genericsMacro } from "./generics.js";
import { Expr } from "../../syntax-objects/expr.js";
import { htmlMacro } from "./html/html.js";
import { listReader } from "./list.js";
import { whitespaceReader } from "./whitespace.js";
import { identifierReader } from "./identifier.js";

const MACROS = [
  listReader,
  whitespaceReader,
  objectLiteralMacro,
  arrayLiteralMacro,
  mapLiteralMacro,
  intMacro,
  floatMacro,
  scientificENotationMacro,
  stringMacro,
  comment,
  booleanMacro,
  genericsMacro,
  htmlMacro,
  identifierReader,
];

export const FAST_MACROS = [
  listReader,
  whitespaceReader,
  objectLiteralMacro,
  arrayLiteralMacro,
  mapLiteralMacro,
  intMacro,
  floatMacro,
  stringMacro,
  comment,
  booleanMacro,
  genericsMacro,
  identifierReader,
];

export const getReaderMacroForToken = (
  token: Token,
  prev?: Expr,
  /** Next char */
  next?: string,
  macros = MACROS
): ReaderMacro["macro"] | undefined =>
  macros.find((m) => m.match(token, prev, next))?.macro;
