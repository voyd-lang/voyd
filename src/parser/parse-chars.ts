import { Expr } from "../syntax-objects/expr.js";
import { List } from "../syntax-objects/list.js";
import { CharStream } from "./char-stream.js";
import { Lexer } from "./lexer.js";
import { getReaderMacroForToken } from "./reader-macros/index.js";
import { Token } from "./token.js";

export type ParseCharsOpts = {
  nested?: boolean;
  terminator?: string;
  parent?: Expr;
  lexer?: Lexer;
};

export const parseChars = (
  file: CharStream,
  opts: ParseCharsOpts = {}
): List => {
  const lexer = opts.lexer ?? new Lexer();
  const list = new List({
    location: file.currentSourceLocation(),
    parent: opts.parent,
  });

  while (file.hasCharacters) {
    const token = lexer.tokenize(file);

    if (token.value === opts.terminator) {
      break;
    }

    if (processWithReaderMacro(token, list.last(), file, list, lexer)) {
      continue;
    }
  }

  list?.setEndLocationToStartOf(file.currentSourceLocation());
  return opts.nested ? list : new List(["ast", list]);
};

/** Returns true if token was matched with and processed by a macro  */
const processWithReaderMacro = (
  token: Token,
  prev: Expr | undefined,
  file: CharStream,
  list: List,
  lexer: Lexer,
) => {
  const readerMacro = getReaderMacroForToken(token, prev, file.next);
  if (!readerMacro) return undefined;

  const result = readerMacro(file, {
    token,
    reader: (file, terminator) =>
      parseChars(file, {
        nested: true,
        terminator,
        lexer,
      }),
  });

  if (result) list.push(result);
  return true;
};
