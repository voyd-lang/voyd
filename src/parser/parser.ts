import { List } from "../syntax-objects/index.js";
import { CharStream } from "./char-stream.js";
import { parseChars } from "./parse-chars.js";
import { ReaderMacro } from "./reader-macros/types.js";
import { expandSyntaxMacros } from "./syntax-macros/index.js";

export const parse = (text: string, filePath?: string): List => {
  const chars = new CharStream(text, filePath ?? "raw");
  const rawAst = parseChars(chars);
  return expandSyntaxMacros(rawAst);
};
