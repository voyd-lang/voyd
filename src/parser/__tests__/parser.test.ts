import { CharStream } from "../char-stream.js";
import { parseChars } from "../parse-chars.js";
import { parse } from "../parser.js";
import { BENCH_FILE, BENCH_FILE_SM } from "./fixtures/benchmark-file.js";
import { voydFile, voydFileWithGenerics } from "./fixtures/voyd-file.js";
import { test } from "vitest";

test("parser can parse a file into a syntax expanded ast", async (t) => {
  t.expect(parse(voydFile)).toMatchSnapshot();
});

test("parser supports generics", async (t) => {
  t.expect(parse(voydFileWithGenerics)).toMatchSnapshot();
});

test("parseChars throughput budget (>= 2 MB/s)", (t) => {
  let sink = 0; // prevent dead-code elimination

  const opsPerSec = measureOpsPerSec(() => {
    const chars = new CharStream(BENCH_FILE_SM, "raw");
    return (sink ^= parseChars(chars) ? 1 : 0);
  });

  const bytesPerOp = Buffer.byteLength(BENCH_FILE_SM, "utf8");
  const mbPerSec = (opsPerSec * bytesPerOp) / 1e6;
  logPerf(mbPerSec, opsPerSec, bytesPerOp);
  t.expect(mbPerSec).toBeGreaterThanOrEqual(0.4);
});

test("parser throughput budget (>= 0.5 MB/s)", (t) => {
  let sink = 0; // prevent dead-code elimination
  const opsPerSec = measureOpsPerSec(
    () => (sink ^= parse(BENCH_FILE_SM) ? 1 : 0)
  );
  const bytesPerOp = Buffer.byteLength(BENCH_FILE_SM, "utf8");
  const mbPerSec = (opsPerSec * bytesPerOp) / 1e6;
  logPerf(mbPerSec, opsPerSec, bytesPerOp);
  t.expect(mbPerSec).toBeGreaterThanOrEqual(0.2);
});

function measureOpsPerSec(fn: () => void, warmupMs = 30, runMs = 200) {
  const until = (t: number) => {
    while (performance.now() < t) fn();
  };
  until(performance.now() + warmupMs);
  let n = 0;
  const end = performance.now() + runMs;
  const start = performance.now();
  while (performance.now() < end) {
    fn();
    n++;
  }
  const total = performance.now() - start;
  return (n / total) * 1000;
}

const logPerf = (mbPerSec: number, opsPerSec: number, bytesPerOp: number) => {
  const mbps = mbPerSec.toLocaleString();
  const ops = opsPerSec.toLocaleString();
  const mb = (bytesPerOp / 1e6).toLocaleString();
  const str = `Achieved total throughput of ${mbps} MB/s (${ops} OP/s * ${mb} MB)`;
  console.log(str);
};
