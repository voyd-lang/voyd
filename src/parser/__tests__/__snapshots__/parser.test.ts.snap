// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parser can parse a file into a syntax expanded ast 1`] = `
[
  "ast",
  [
    "use",
    [
      "::",
      [
        "::",
        "std",
        "macros",
      ],
      "all",
    ],
  ],
  [
    "use",
    [
      "::",
      [
        "::",
        "std",
        "io",
      ],
      [
        "object",
        "read",
        [
          ":",
          "write",
          "io_write",
        ],
      ],
    ],
  ],
  [
    "fn",
    [
      "->",
      [
        "fib",
        [
          ":",
          "n",
          "i32",
        ],
      ],
      "i32",
    ],
    [
      "block",
      [
        "if",
        [
          "<=",
          "n",
          1,
        ],
        [
          ":",
          "then",
          [
            "block",
            "n",
          ],
        ],
        [
          ":",
          "else",
          [
            "block",
            [
              "+",
              [
                "fib",
                [
                  "-",
                  "n",
                  1,
                ],
              ],
              [
                "fib",
                [
                  "-",
                  "n",
                  2,
                ],
              ],
            ],
          ],
        ],
      ],
    ],
  ],
  [
    "macro_let",
    [
      "=",
      "extract_parameters",
      [
        "=>",
        [
          "definitions",
        ],
        [
          "block",
          [
            "concat",
            [
              "\`",
              "parameters",
            ],
            [
              "slice",
              "definitions",
              1,
            ],
          ],
        ],
      ],
    ],
  ],
  [
    "if",
    [
      ">",
      "x",
      10,
    ],
    [
      ":",
      "then",
      [
        "block",
        10,
      ],
    ],
    [
      ":",
      "else",
      [
        "block",
        20,
      ],
    ],
  ],
  [
    "if",
    [
      "<",
      "x",
      {
        "type": "f64",
        "value": 0,
      },
    ],
    [
      ":",
      "then",
      -1,
    ],
    [
      ":",
      "elif",
      [
        "<",
        "x",
        {
          "type": "f64",
          "value": 2,
        },
      ],
    ],
    [
      ":",
      "then",
      0,
    ],
    [
      ":",
      "else",
      2,
    ],
  ],
  [
    "reduce",
    "array",
    0,
    1,
    2,
    [
      ":",
      "hey",
      [
        "=>",
        [],
        [
          "block",
          [
            "log",
            "val",
          ],
          [
            "+",
            "acc",
            "val",
          ],
        ],
      ],
    ],
    [
      ":",
      "with",
      [
        "=>",
        [],
        0,
      ],
    ],
  ],
  10,
  [
    "+",
    3,
  ],
  [
    "mul",
    [
      "&",
      "self",
    ],
    [
      ":",
      "other",
      [
        "Vec",
        [
          "generics",
          "T",
        ],
      ],
    ],
  ],
  [
    "mul",
    [
      "&",
      "self",
    ],
    [
      ":",
      "other",
      [
        "Vec",
        [
          "generics",
          "T",
        ],
      ],
    ],
  ],
  [
    "let",
    [
      "=",
      "a",
      [
        "+",
        [
          "reduce",
          "array",
          0,
          [
            "=>",
            [
              "acc",
              "val",
            ],
            [
              "block",
              [
                "+",
                "acc",
                "val",
              ],
            ],
          ],
        ],
        [
          "*",
          10,
          3,
        ],
      ],
    ],
  ],
  [
    "let",
    [
      "=",
      "x",
      [
        "my_func",
        [
          "add",
          1,
          2,
        ],
        [
          "=>",
          [],
          [
            "block",
            [
              "hello",
            ],
          ],
        ],
        [
          "+",
          3,
          4,
        ],
      ],
    ],
  ],
  [
    "let",
    [
      "=",
      "vec",
      [
        "object",
        [
          ":",
          "a",
          [
            "hey",
            "there",
          ],
        ],
        [
          ":",
          "b",
          2,
        ],
      ],
    ],
  ],
  [
    "closure_param_test",
    1,
    [
      "=>",
      [],
      "a",
    ],
    3,
    [
      "=>",
      [],
      [
        "block",
        [
          "hey",
          "there",
        ],
      ],
    ],
    4,
    [
      "=>",
      [],
      5,
    ],
    [
      "=>",
      [],
      [
        "block",
        6,
      ],
    ],
    [
      "=>",
      [],
      [
        "block",
        7,
      ],
    ],
    8,
  ],
  [
    "let",
    [
      "=",
      [
        "tuple",
        "x",
        "y",
      ],
      [
        "tuple",
        1,
        2,
      ],
    ],
  ],
  [
    "+",
    [
      "Array",
      [
        "generics",
        "Hey",
        "There",
      ],
      1,
      2,
      3,
    ],
    3,
  ],
  [
    "obj",
    [
      "Test",
      [
        "generics",
        "T",
      ],
    ],
    [
      "object",
      [
        ":",
        "c",
        "i32",
      ],
    ],
  ],
  [
    "match",
    "x",
    [
      ":",
      [
        "Some",
        [
          "generics",
          "i32",
        ],
      ],
      [
        "Some",
        [
          "generics",
          "i32",
        ],
        [
          "object",
          [
            ":",
            "value",
            [
              "+",
              [
                "value",
                "x",
              ],
              1,
            ],
          ],
        ],
      ],
    ],
    [
      ":",
      "None",
      [
        "None",
        [
          "object",
        ],
      ],
    ],
  ],
  [
    "match",
    [
      "match",
      [
        "pop",
        "arr",
      ],
      "val",
      [
        ":",
        [
          "Some",
          [
            "generics",
            [
              "Optional",
              [
                "generics",
                "i32",
              ],
            ],
          ],
        ],
        [
          "block",
          [
            "value",
            "val",
          ],
        ],
      ],
      [
        ":",
        "None",
        [
          "block",
          [
            "None",
            [
              "object",
            ],
          ],
        ],
      ],
    ],
    "v",
    [
      ":",
      [
        "Some",
        [
          "generics",
          "i32",
        ],
      ],
      [
        "block",
        [
          "value",
          "v",
        ],
      ],
    ],
    [
      ":",
      "None",
      [
        "block",
        -1,
      ],
    ],
  ],
  [
    "match",
    [
      "match",
      [
        "pop",
        "arr",
      ],
      "val",
      [
        ":",
        [
          "Some",
          [
            "generics",
            [
              "Optional",
              [
                "generics",
                "i32",
              ],
            ],
          ],
        ],
        [
          "block",
          [
            "value",
            "val",
          ],
        ],
      ],
      [
        ":",
        "None",
        [
          "block",
          [
            "None",
            [
              "object",
            ],
          ],
        ],
      ],
    ],
    "v",
    [
      ":",
      [
        "Some",
        [
          "generics",
          "i32",
        ],
      ],
      [
        "block",
        [
          "value",
          "v",
        ],
      ],
    ],
    [
      ":",
      "None",
      [
        "block",
        -1,
      ],
    ],
  ],
  [
    "fn",
    [
      "->",
      [
        "test",
        [
          "generics",
          "T",
        ],
        [
          ":",
          "a",
          1,
        ],
      ],
      "i32",
    ],
  ],
  [
    "fn",
    [
      "main",
    ],
    [
      "block",
      [
        "let",
        [
          "=",
          "a",
          [
            "+",
            [
              "...",
              [
                "hey",
                "test",
              ],
            ],
            [
              "now",
              [
                "&",
                "other",
              ],
            ],
          ],
        ],
      ],
      [
        "let",
        [
          "=",
          "x",
          [
            "+",
            10,
            [
              "block",
              [
                "+",
                20,
                30,
              ],
            ],
          ],
        ],
      ],
      [
        "let",
        [
          "=",
          "y",
          [
            "if",
            [
              ">",
              "x",
              10,
            ],
            [
              ":",
              "then",
              [
                "block",
                10,
              ],
            ],
            [
              ":",
              "else",
              [
                "block",
                20,
              ],
            ],
          ],
        ],
      ],
      [
        "let",
        [
          "=",
          "n",
          [
            "block",
            [
              "if",
              [
                ">",
                [
                  "len",
                  "args",
                ],
                1,
              ],
              [
                ":",
                "then",
                [
                  "block",
                  [
                    "log",
                    "console",
                    [
                      "new_string",
                      [
                        "object",
                        "ObjectLiteral-840",
                        [
                          [
                            "from",
                            [
                              "FixedArray",
                              72,
                              101,
                              121,
                              32,
                              116,
                              104,
                              101,
                              114,
                              101,
                              33,
                            ],
                          ],
                        ],
                      ],
                    ],
                  ],
                  [
                    "unwrap",
                    [
                      "parseInt",
                      [
                        "at",
                        "args",
                        1,
                      ],
                    ],
                  ],
                ],
              ],
              [
                ":",
                "else",
                [
                  "block",
                  10,
                ],
              ],
            ],
          ],
        ],
      ],
      [
        "let",
        [
          "=",
          "x2",
          10,
        ],
      ],
      [
        "let",
        [
          "=",
          "z",
          [
            "nothing",
          ],
        ],
      ],
      [
        "let",
        [
          "=",
          "a",
          [
            "boop",
            "hello",
            1,
          ],
        ],
      ],
      [
        "let",
        [
          "=",
          "test_spacing",
          [
            "fib",
            "n",
          ],
        ],
      ],
      [
        "let",
        [
          "=",
          "result",
          [
            "fib",
            "n",
          ],
        ],
      ],
      [
        "let",
        [
          "=",
          "x",
          [
            "&",
            "hey",
          ],
        ],
      ],
      [
        "$",
        "hey",
      ],
      [
        "$@",
        [
          "hey",
        ],
      ],
      [
        "$",
        [
          "hey",
        ],
      ],
      [
        "$",
        [
          "extract",
          "equals_expr",
          2,
        ],
      ],
      [
        "block",
        [
          "$",
          "body",
        ],
      ],
      [
        "+",
        "x",
        5,
      ],
      [
        "+",
        [
          "+",
          "x",
          "y",
        ],
        10,
      ],
      [
        "+",
        [
          "*",
          "x",
          "y",
        ],
        10,
      ],
      [
        "x",
      ],
      "x",
      [
        "let",
        [
          "=",
          "vec",
          [
            "object",
            [
              ":",
              "x",
              10,
            ],
            [
              ":",
              "y",
              [
                "Point",
                [
                  "object",
                  [
                    ":",
                    "x",
                    10,
                  ],
                  [
                    ":",
                    "y",
                    20,
                  ],
                ],
              ],
            ],
            [
              ":",
              "z",
              [
                "object",
                [
                  ":",
                  "a",
                  10,
                ],
                [
                  ":",
                  "b",
                  20,
                ],
              ],
            ],
          ],
        ],
      ],
      [
        "pub",
        "fn",
        [
          "main",
        ],
        [
          "block",
          [
            "html",
            [
              "div",
              [
                ":",
                "attributes",
                [
                  "dict",
                  "array",
                ],
              ],
              [
                ":",
                "children",
                [
                  "p",
                  [
                    ":",
                    "attributes",
                    [
                      "dict",
                      "array",
                    ],
                  ],
                  [
                    ":",
                    "children",
                    [
                      "array",
                      [
                        "new_string",
                        [
                          "object",
                          "ObjectLiteral-1125",
                          [
                            [
                              "from",
                              [
                                "FixedArray",
                                72,
                                101,
                                108,
                                108,
                                111,
                                32,
                                116,
                                104,
                                101,
                                114,
                                101,
                                33,
                                33,
                                10,
                                32,
                                32,
                                32,
                                32,
                                32,
                                32,
                              ],
                            ],
                          ],
                        ],
                      ],
                    ],
                  ],
                  [
                    "p",
                    [
                      ":",
                      "attributes",
                      [
                        "dict",
                        "array",
                      ],
                    ],
                    [
                      ":",
                      "children",
                      [
                        "array",
                        [
                          "new_string",
                          [
                            "object",
                            "ObjectLiteral-1173",
                            [
                              [
                                "from",
                                [
                                  "FixedArray",
                                  72,
                                  111,
                                  119,
                                  32,
                                  97,
                                  114,
                                  101,
                                  32,
                                  121,
                                  111,
                                  117,
                                  32,
                                  116,
                                  111,
                                  100,
                                  97,
                                  121,
                                  10,
                                  32,
                                  32,
                                  32,
                                  32,
                                  32,
                                  32,
                                ],
                              ],
                            ],
                          ],
                        ],
                      ],
                    ],
                  ],
                ],
              ],
            ],
          ],
        ],
      ],
    ],
  ],
]
`;

exports[`parser supports generics 1`] = `
[
  "ast",
  [
    "use",
    [
      "::",
      "std",
      "all",
    ],
  ],
  [
    "type",
    [
      "=",
      "DsArrayi32",
      [
        "DsArray",
        [
          "generics",
          "i32",
        ],
      ],
    ],
  ],
  [
    "pub",
    "fn",
    [
      "main",
    ],
    [
      "block",
      [
        "let",
        [
          "=",
          "arr",
          [
            "new_fixed_array",
            [
              "generics",
              "i32",
            ],
            10,
          ],
        ],
      ],
      [
        "set",
        [
          "generics",
          "i32",
        ],
        "arr",
        0,
        1,
      ],
      [
        "get",
        [
          "generics",
          "i32",
        ],
        "arr",
        0,
      ],
    ],
  ],
]
`;
